File: hard/42. Trapping Rain Water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function (height) {
    let left = 0;
    let right = height.length - 1;
    let ans = 0;
    let leftMax = 0;
    let rightMax = 0;

    while (left < right) {
        if (height[left] < height[right]) {
            leftMax = Math.max(leftMax, height[left]);
            ans += leftMax - height[left];
            ++left;
        } else {
            rightMax = Math.max(rightMax, height[right]);
            ans += rightMax - height[right];
            --right;
        }
    }

    return ans;
};

File: python/dynamic programming/42. Trapping Rain Water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        i = 0
        k = len(height) - 1

        leftMax = 0
        rightMax = 0

        water = 0

        while (i < len(height) and k >= 0) and (i < k):
            leftMax = max(leftMax, height[i])
            rightMax = max(rightMax, height[k])

            if leftMax < rightMax:
                i += 1
                if height[i] < leftMax:
                    water += leftMax - height[i]
            else:
                if height[k] < rightMax:
                    water += rightMax - height[k]
                k -= 1

            print(leftMax, rightMax)

        return water
